data_structures/priority_queue.cpp

```diff
@@ -0,0 +1,154 @@
+/**
+ * @file
+ * @brief Implemetation of Max Priority Queue uning max-heap
+ * More about it: https://en.wikipedia.org/wiki/Priority_queue
+ * @details
+ * In computer science, a priority queue is an abstract data-type similar to a 
+ * regular queue or stack data structure in which each element additionally has a 
+ * priority associated with it. In a priority queue, an element with high priority 
+ * is served before an element with low priority.
+ * Though being a relatively new data structure it has its use cases in lots of fields:
+ * Data compression[Huffman codes], Graph Searching[Dijkstra's Algorithm], 
+ * Artificial Intelligence[A* seach], etc.
+ * Insertion - O(log(N))
+ * Deletion - O(log(N))
+ * Max - O(1)
+ * All the method names have been directly taken from PQ implementation of Java
+ * @author [Puneet Tripathi](https://github.com/PuneetTri)
+ */
+
+#include<iostream>
+using namespace std;
+
+// Class 
+class PQ{
+    private:
+        int capacity; // max capacity of PQ
+        int *arr = NULL; // Pointer to the array of the PQ
+        int N; // current size of PQ or total elements currently in PQ
+        
+        // Helper function to swap elements
+        void swap(int *a, int *b) {
+            int temp = *a;
+            *a = *b;
+            *b = temp;
+        }
+
+        /* Helper function for the max-heap implementation of PQ
+        *  Function would help promte child node to their correct 
+        *  position
+        */
+        void swim(int pos) {
+            // Check if parent is smaller than child node
+            while(pos > 1 && (arr[pos/2] < arr[pos])) {
+                // In such case swap value of child with parent
+                swap(&arr[pos], &arr[pos/2]);
+                pos = pos/2; // Jump to position of parent node
+            }
+            // Promotion of child node will go on until it becomes smaller than the parent
+        }
+
+        /* Helper function for the max-heap implementation of PQ
+        *  Function would help demote parent node to their correct 
+        *  position
+        */
+        void sink(int pos) {
+            // Check if node's position is that of parent node
+            while(2*pos <= N) {
+                int current = 2*pos; // Jump to the positon of child node
+                // Compare both the children for the greater one
+                if(current < N && arr[current] < arr[current+1]) current++;
+                // If the parent node is greater, sink operation is complete. Break the loop
+                if(arr[pos] >= arr[current]) break; 
+                swap(&arr[pos], &arr[current]); // If not exchange the value of parent with child
+                pos = current; // Exchange parent position to child position in the array
+            }
+        }
+    public:
+        PQ() {
```

Comment:
you can remove the constructor PQ() and defaulting the argument of PQ(int max) with. this way redundant code could be avoided. Another possibility would be to call PQ(int max) from PQ() with the desired parameter.